<?xml version="1.0" encoding="utf-8"?>

<!-- 
#  
# Title:
#  Open Source Lunar Rover (OSLR) Model - Wheels
#
# Author(s):
#   andyjak
#
# Version:
#   0.0.1, 08/2023
#
# Purpose:
#   Contains the steering and drive of the wheels.
#
# Notes:
#   Include this file in the urdf file.
#
# Test setup:
#   - ROS Noetic
#   - Ubuntu, 20.04 LTS
#   - Python 3.8.10
#
# References:
#   ...
#
-->

<robot
  xmlns:xacro="http://ros.org/wiki/xacro"
  name="oslr_wheels">
  
  <!-- ********************************** PARAMETER VALUES ********************************** -->
  <xacro:property name="steer_angle_lim_upper" value="${yaml_dict['steer_angle_lim_upper']}" />
  <xacro:property name="steer_angle_lim_lower" value="${yaml_dict['steer_angle_lim_lower']}" />
  <xacro:property name="steer_effort_lim" value="${yaml_dict['steer_effort_lim']}" />
  <xacro:property name="steer_velocity_lim" value="${yaml_dict['steer_velocity_lim']}" />
  <xacro:property name="drive_effort_lim" value="${yaml_dict['drive_effort_lim']}" />
  <xacro:property name="drive_velocity_lim" value="${yaml_dict['drive_velocity_lim']}" />
  <xacro:property name="friction_coeff1_wheel" value="${yaml_dict['friction_coeff1_wheel']}" />
  <xacro:property name="friction_coeff2_wheel" value="${yaml_dict['friction_coeff2_wheel']}" />
  <xacro:property name="stiffness_wheel" value="${yaml_dict['stiffness_wheel']}" />
  <xacro:property name="damping_wheel" value="${yaml_dict['damping_wheel']}" />
  <xacro:property name="friction_coeff1_corner" value="${yaml_dict['friction_coeff1_corner']}" />
  <xacro:property name="friction_coeff2_corner" value="${yaml_dict['friction_coeff2_corner']}" />
  <xacro:property name="stiffness_corner" value="${yaml_dict['stiffness_corner']}" />
  <xacro:property name="damping_corner" value="${yaml_dict['damping_corner']}" />
  
  <!-- ********************************** GEOMETRY ********************************** -->
  <link
    name="front_left_corner">
    <inertial>
      <origin
        xyz="-0.00021559 -0.058703 -0.078963"
        rpy="0 0 0" />
      <mass
        value="0.17725" />
      <inertia
        ixx="0.00012121"
        ixy="8.0199E-08"
        ixz="1.5423E-07"
        iyy="7.6277E-05"
        iyz="7.7454E-08"
        izz="0.00011786" />
    </inertial>
    <visual>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://oslr/models/oslr/meshes/front_left_corner.STL" />
      </geometry>
      <material
        name="">
        <color
          rgba="0.89804 0.91765 0.92941 1" />
      </material>
    </visual>
    <collision>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://oslr/models/oslr/meshes/front_left_corner.STL" />
      </geometry>
    </collision>
  </link>
  <joint
    name="front_left_corner_joint"
    type="revolute">
    <origin
      xyz="0.215380376567892 0.00552677512849714 0.350915322468129"
      rpy="1.57079631098936 0 0" />
    <parent
      link="left_bogie" />
    <child
      link="front_left_corner" />
    <axis
      xyz="0 0 -0.999999999999988" />
    <limit
      lower="${steer_angle_lim_lower}"
      upper="${steer_angle_lim_upper}"
      effort="${steer_effort_lim}"
      velocity="${steer_velocity_lim}" />
  </joint>
  <link
    name="front_left_wheel">
    <inertial>
      <origin
        xyz="4.9366E-10 -1.3882E-10 0.0071368"
        rpy="0 0 0" />
      <mass
        value="0.61431" />
      <inertia
        ixx="0.0011197"
        ixy="5.7673E-08"
        ixz="-3.3832E-16"
        iyy="0.0011199"
        iyz="5.6789E-17"
        izz="0.001892" />
    </inertial>
    <visual>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://oslr/models/oslr/meshes/front_left_wheel.STL" />
      </geometry>
      <material
        name="">
        <color
          rgba="0.62745 0.62745 0.62745 1" />
      </material>
    </visual>
    <collision>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://oslr/models/oslr/meshes/front_left_wheel.STL" />
      </geometry>
    </collision>
  </link>
  <joint
    name="front_left_drive_joint"
    type="continuous">
    <origin
      xyz="-0.00017180884072221 -4.3244793997782E-05 -0.128892240205177"
      rpy="-1.57079632679486 0 0" />
    <parent
      link="front_left_corner" />
    <child
      link="front_left_wheel" />
    <axis
      xyz="0 0 1" />
    <limit
      effort="${drive_effort_lim}"
      velocity="${drive_velocity_lim}" />
  </joint>
  <link
    name="middle_left_wheel">
    <inertial>
      <origin
        xyz="4.4851E-10 2.7786E-10 0.0071368"
        rpy="0 0 0" />
      <mass
        value="0.61431" />
      <inertia
        ixx="0.0011197"
        ixy="5.2219E-08"
        ixz="-1.5331E-16"
        iyy="0.0011199"
        iyz="-3.1316E-17"
        izz="0.001892" />
    </inertial>
    <visual>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://oslr/models/oslr/meshes/middle_left_wheel.STL" />
      </geometry>
      <material
        name="">
        <color
          rgba="0.62745 0.62745 0.62745 1" />
      </material>
    </visual>
    <collision>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://oslr/models/oslr/meshes/middle_left_wheel.STL" />
      </geometry>
    </collision>
  </link>
  <joint
    name="middle_left_drive_joint"
    type="continuous">
    <origin
      xyz="-0.0639545058516444 0.103117089777728 0.436448751635148"
      rpy="0 0 0" />
    <parent
      link="left_bogie" />
    <child
      link="middle_left_wheel" />
    <axis
      xyz="0 0 1" />
    <limit
      effort="${drive_effort_lim}"
      velocity="${drive_velocity_lim}" />
  </joint>
  <link
    name="back_left_corner">
    <inertial>
      <origin
        xyz="0.00012803 -0.058617 -0.078963"
        rpy="0 0 0" />
      <mass
        value="0.17725" />
      <inertia
        ixx="0.00012121"
        ixy="8.0199E-08"
        ixz="1.5423E-07"
        iyy="7.6277E-05"
        iyz="7.7454E-08"
        izz="0.00011786" />
    </inertial>
    <visual>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://oslr/models/oslr/meshes/back_left_corner.STL" />
      </geometry>
      <material
        name="">
        <color
          rgba="0.89804 0.91765 0.92941 1" />
      </material>
    </visual>
    <collision>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://oslr/models/oslr/meshes/back_left_corner.STL" />
      </geometry>
    </collision>
  </link>
  <joint
    name="back_left_corner_joint"
    type="revolute">
    <origin
      xyz="0.095997370210931 -0.186901909356396 0.0342697447940012"
      rpy="-1.5707963267949 0 1.61170771758594" />
    <parent
      link="left_rocker" />
    <child
      link="back_left_corner" />
    <axis
      xyz="0 0 -1" />
    <limit
      lower="${steer_angle_lim_lower}"
      upper="${steer_angle_lim_upper}"
      effort="${steer_effort_lim}"
      velocity="${steer_velocity_lim}" />
  </joint>
  <link
    name="back_left_wheel">
    <inertial>
      <origin
        xyz="4.3795E-10 6.9914E-10 0.0071368"
        rpy="0 0 0" />
      <mass
        value="0.61431" />
      <inertia
        ixx="0.0011197"
        ixy="5.7673E-08"
        ixz="-1.2405E-16"
        iyy="0.0011199"
        iyz="1.2035E-16"
        izz="0.001892" />
    </inertial>
    <visual>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://oslr/models/oslr/meshes/back_left_wheel.STL" />
      </geometry>
      <material
        name="">
        <color
          rgba="0.62745 0.62745 0.62745 1" />
      </material>
    </visual>
    <collision>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://oslr/models/oslr/meshes/back_left_wheel.STL" />
      </geometry>
    </collision>
  </link>
  <joint
    name="back_left_drive_joint"
    type="continuous">
    <origin
      xyz="0.000171808840937315 4.32447939933134E-05 -0.128892240205023"
      rpy="-1.57079632679486 0 0" />
    <parent
      link="back_left_corner" />
    <child
      link="back_left_wheel" />
    <axis
      xyz="0 0 1" />
    <limit
      effort="${drive_effort_lim}"
      velocity="${drive_velocity_lim}" />
  </joint>
  <link
    name="front_right_corner">
    <inertial>
      <origin
        xyz="-0.00027466 0.05974 -0.078145"
        rpy="0 0 0" />
      <mass
        value="0.1709" />
      <inertia
        ixx="0.00011992"
        ixy="8.0199E-08"
        ixz="-1.5237E-07"
        iyy="7.3812E-05"
        iyz="-7.749E-08"
        izz="0.00011672" />
    </inertial>
    <visual>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://oslr/models/oslr/meshes/front_right_corner.STL" />
      </geometry>
      <material
        name="">
        <color
          rgba="0.89804 0.91765 0.92941 1" />
      </material>
    </visual>
    <collision>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://oslr/models/oslr/meshes/front_right_corner.STL" />
      </geometry>
    </collision>
  </link>
  <joint
    name="front_right_corner_joint"
    type="revolute">
    <origin
      xyz="0.213307849599925 0.00937420353549007 -0.350130330857906"
      rpy="1.57079655183463 0 0" />
    <parent
      link="right_bogie" />
    <child
      link="front_right_corner" />
    <axis
      xyz="0 0 -1" />
    <limit
      lower="${steer_angle_lim_lower}"
      upper="${steer_angle_lim_upper}"
      effort="${steer_effort_lim}"
      velocity="${steer_velocity_lim}" />
  </joint>
  <link
    name="front_right_wheel">
    <inertial>
      <origin
        xyz="5.237E-11 -5.0691E-11 -0.0071368"
        rpy="0 0 0" />
      <mass
        value="0.61431" />
      <inertia
        ixx="0.0011197"
        ixy="-5.7673E-08"
        ixz="-1.312E-16"
        iyy="0.0011199"
        iyz="-5.7322E-17"
        izz="0.001892" />
    </inertial>
    <visual>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://oslr/models/oslr/meshes/front_right_wheel.STL" />
      </geometry>
      <material
        name="">
        <color
          rgba="0.62745 0.62745 0.62745 1" />
      </material>
    </visual>
    <collision>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://oslr/models/oslr/meshes/front_right_wheel.STL" />
      </geometry>
    </collision>
  </link>
  <joint
    name="front_right_drive_joint"
    type="continuous">
    <origin
      xyz="-0.000171808840730647 -4.32447939935632E-05 -0.128892240205185"
      rpy="-1.57079632679494 0 0" />
    <parent
      link="front_right_corner" />
    <child
      link="front_right_wheel" />
    <axis
      xyz="0 0 1" />
    <limit
      effort="${drive_effort_lim}"
      velocity="${drive_velocity_lim}" />
  </joint>
  <link
    name="middle_right_wheel">
    <inertial>
      <origin
        xyz="4.161E-11 3.743E-11 -0.0071368"
        rpy="0 0 0" />
      <mass
        value="0.61431" />
      <inertia
        ixx="0.0011197"
        ixy="-5.2219E-08"
        ixz="-1.6248E-16"
        iyy="0.0011199"
        iyz="3.137E-17"
        izz="0.001892" />
    </inertial>
    <visual>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://oslr/models/oslr/meshes/middle_right_wheel.STL" />
      </geometry>
      <material
        name="">
        <color
          rgba="0.62745 0.62745 0.62745 1" />
      </material>
    </visual>
    <collision>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://oslr/models/oslr/meshes/middle_right_wheel.STL" />
      </geometry>
    </collision>
  </link>
  <joint
    name="middle_right_drive_joint"
    type="continuous">
    <origin
      xyz="-0.0660270326936359 0.106964536461386 -0.435750229193693"
      rpy="0 0 0" />
    <parent
      link="right_bogie" />
    <child
      link="middle_right_wheel" />
    <axis
      xyz="0 0 1" />
    <limit
      effort="${drive_effort_lim}"
      velocity="${drive_velocity_lim}" />
  </joint>
  <link
    name="back_right_corner">
    <inertial>
      <origin
        xyz="6.896E-05 0.059827 -0.078145"
        rpy="0 0 0" />
      <mass
        value="0.1709" />
      <inertia
        ixx="0.00011992"
        ixy="8.0199E-08"
        ixz="-1.5237E-07"
        iyy="7.3812E-05"
        iyz="-7.749E-08"
        izz="0.00011672" />
    </inertial>
    <visual>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://oslr/models/oslr/meshes/back_right_corner.STL" />
      </geometry>
      <material
        name="">
        <color
          rgba="0.89804 0.91765 0.92941 1" />
      </material>
    </visual>
    <collision>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://oslr/models/oslr/meshes/back_right_corner.STL" />
      </geometry>
    </collision>
  </link>
  <joint
    name="back_right_corner_joint"
    type="revolute">
    <origin
      xyz="-0.187875925745586 0.0940768587729567 0.0343074274445147"
      rpy="1.57079632679489 0 -0.0306626789685753" />
    <parent
      link="right_rocker" />
    <child
      link="back_right_corner" />
    <axis
      xyz="0 0 -1" />
    <limit
      lower="${steer_angle_lim_lower}"
      upper="${steer_angle_lim_upper}"
      effort="${steer_effort_lim}"
      velocity="${steer_velocity_lim}" />
  </joint>
  <link
    name="back_right_wheel">
    <inertial>
      <origin
        xyz="2.8141E-11 8.3344E-11 -0.0071368"
        rpy="0 0 0" />
      <mass
        value="0.61431" />
      <inertia
        ixx="0.0011197"
        ixy="-5.7673E-08"
        ixz="-1.5585E-16"
        iyy="0.0011199"
        iyz="-1.1423E-17"
        izz="0.001892" />
    </inertial>
    <visual>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://oslr/models/oslr/meshes/back_right_wheel.STL" />
      </geometry>
      <material
        name="">
        <color
          rgba="0.62745 0.62745 0.62745 1" />
      </material>
    </visual>
    <collision>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://oslr/models/oslr/meshes/back_right_wheel.STL" />
      </geometry>
    </collision>
  </link>
  <joint
    name="back_right_drive_joint"
    type="continuous">
    <origin
      xyz="0.000171808840730536 4.32447939971159E-05 -0.128892240205185"
      rpy="-1.57079632679494 0 0" />
    <parent
      link="back_right_corner" />
    <child
      link="back_right_wheel" />
    <axis
      xyz="0 0 1" />
    <limit
      effort="${drive_effort_lim}"
      velocity="${drive_velocity_lim}" />
  </joint>

  <!-- ********************************** GAZEBO PROPERTIES ********************************** -->
  <gazebo reference="front_left_wheel">
      <mu1>${friction_coeff1_wheel}</mu1>
      <mu2>${friction_coeff2_wheel}</mu2>
      <kp>${stiffness_wheel}</kp>
      <kd>${damping_wheel}</kd>
      <material>Gazebo/Grey</material>
  </gazebo>
  <gazebo reference="front_right_wheel">
      <mu1>${friction_coeff1_wheel}</mu1>
      <mu2>${friction_coeff2_wheel}</mu2>
      <kp>${stiffness_wheel}</kp>
      <kd>${damping_wheel}</kd>
      <material>Gazebo/Grey</material>
  </gazebo>
  <gazebo reference="middle_left_wheel">
      <mu1>${friction_coeff1_wheel}</mu1>
      <mu2>${friction_coeff2_wheel}</mu2>
      <kp>${stiffness_wheel}</kp>
      <kd>${damping_wheel}</kd>
      <material>Gazebo/Grey</material>
  </gazebo>
  <gazebo reference="middle_right_wheel">
      <mu1>${friction_coeff1_wheel}</mu1>
      <mu2>${friction_coeff2_wheel}</mu2>
      <kp>${stiffness_wheel}</kp>
      <kd>${damping_wheel}</kd>
      <material>Gazebo/Grey</material>
  </gazebo>
  <gazebo reference="back_left_wheel">
      <mu1>${friction_coeff1_wheel}</mu1>
      <mu2>${friction_coeff2_wheel}</mu2>
      <kp>${stiffness_wheel}</kp>
      <kd>${damping_wheel}</kd>
      <material>Gazebo/Grey</material>
  </gazebo>
  <gazebo reference="back_right_wheel">
      <mu1>${friction_coeff1_wheel}</mu1>
      <mu2>${friction_coeff2_wheel}</mu2>
      <kp>${stiffness_wheel}</kp>
      <kd>${damping_wheel}</kd>
      <material>Gazebo/Grey</material>
  </gazebo>

  <gazebo reference="front_left_corner">
      <mu1>${friction_coeff1_corner}</mu1>
      <mu2>${friction_coeff2_corner}</mu2>
      <kp>${stiffness_corner}</kp>
      <kd>${damping_corner}</kd>
      <material>Gazebo/White</material>
  </gazebo>

  <gazebo reference="front_right_corner">
      <mu1>${friction_coeff1_corner}</mu1>
      <mu2>${friction_coeff2_corner}</mu2>
      <kp>${stiffness_corner}</kp>
      <kd>${damping_corner}</kd>
      <material>Gazebo/White</material>
  </gazebo>

  <gazebo reference="back_left_corner">
      <mu1>${friction_coeff1_corner}</mu1>
      <mu2>${friction_coeff2_corner}</mu2>
      <kp>${stiffness_corner}</kp>
      <kd>${damping_corner}</kd>
      <material>Gazebo/White</material>
  </gazebo>

  <gazebo reference="back_right_corner">
      <mu1>${friction_coeff1_corner}</mu1>
      <mu2>${friction_coeff2_corner}</mu2>
      <kp>${stiffness_corner}</kp>
      <kd>${damping_corner}</kd>
      <material>Gazebo/White</material>
  </gazebo>

  <!-- ********************************** TRANSMISSIONS ********************************** -->
  <transmission name="front_left_wheel_trans">
    <type>transmission_interface/SimpleTransmission</type>
    <actuator name="front_left_wheel_motor">
      <mechanicalReduction>26.9</mechanicalReduction>
    </actuator>
    <joint name="front_left_drive_joint">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
    </joint>
  </transmission>

  <transmission name="front_right_wheel_trans">
    <type>transmission_interface/SimpleTransmission</type>
    <actuator name="front_right_wheel_motor">
      <mechanicalReduction>26.9</mechanicalReduction>
    </actuator>
    <joint name="front_right_drive_joint">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
    </joint>
  </transmission>

  <transmission name="middle_left_wheel_trans">
    <type>transmission_interface/SimpleTransmission</type>
    <actuator name="middle_left_wheel_motor">
      <mechanicalReduction>26.9</mechanicalReduction>
    </actuator>
    <joint name="middle_left_drive_joint">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
    </joint>
  </transmission>

  <transmission name="middle_right_wheel_trans">
    <type>transmission_interface/SimpleTransmission</type>
    <actuator name="middle_right_wheel_motor">
      <mechanicalReduction>26.9</mechanicalReduction>
    </actuator>
    <joint name="middle_right_drive_joint">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
    </joint>
  </transmission>

  <transmission name="back_left_wheel_trans">
    <type>transmission_interface/SimpleTransmission</type>
    <actuator name="back_left_wheel_motor">
      <mechanicalReduction>26.9</mechanicalReduction>
    </actuator>
    <joint name="back_left_drive_joint">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
    </joint>
  </transmission>

  <transmission name="back_right_wheel_trans">
    <type>transmission_interface/SimpleTransmission</type>
    <actuator name="back_right_wheel_motor">
      <mechanicalReduction>26.9</mechanicalReduction>
    </actuator>
    <joint name="back_right_drive_joint">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
    </joint>
  </transmission>

  <transmission name="front_left_corner_trans">
    <type>transmission_interface/SimpleTransmission</type>
    <actuator name="front_left_corner_motor">
      <mechanicalReduction>189</mechanicalReduction>
    </actuator>
    <joint name="front_left_corner_joint">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
    </joint>
  </transmission>

  <transmission name="front_right_corner_trans">
    <type>transmission_interface/SimpleTransmission</type>
    <actuator name="front_right_corner_motor">
      <mechanicalReduction>189</mechanicalReduction>
    </actuator>
    <joint name="front_right_corner_joint">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
    </joint>
  </transmission>

  <transmission name="back_left_corner_trans">
    <type>transmission_interface/SimpleTransmission</type>
    <actuator name="back_left_corner_motor">
      <mechanicalReduction>189</mechanicalReduction>
    </actuator>
    <joint name="back_left_corner_joint">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
    </joint>
  </transmission>

  <transmission name="back_right_corner_trans">
    <type>transmission_interface/SimpleTransmission</type>
    <actuator name="back_right_corner_motor">
      <mechanicalReduction>189</mechanicalReduction>
    </actuator>
    <joint name="back_right_corner_joint">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
    </joint>
  </transmission>

</robot>